lib/moesif_rack/moesif_middleware.rb

```diff
@@ -36,6 +36,60 @@ def initialize app, options = {}
       @config_dict = Hash.new
       @disable_transaction_id = options['disable_transaction_id'] || false
       @log_body = options.fetch('log_body', true)
+      @batch_size = options['batch_size'] || 25
+      @events_queue = Queue.new
+      @event_response_config_etag = nil
+
+      Thread::new do
+        loop do
+          begin
+            if @events_queue.size > 0
+              has_events = true
+            else
+              has_events = false
+            end
+  
+            until has_events == false do 
+              if @events_queue.size > 0
+                batch_events = []
+                until batch_events.size == @batch_size || @events_queue.empty? do 
+                  batch_events << @events_queue.pop
+                  if (batch_events.size == @batch_size) 
```

Comment:
how about this?

Reply:
If statement is inside the until block. We're iterating through the queue and storing it in local array when reading from the queue. Once it matches the batch size, we send events. We also send events if we read events from the queue but it's less than batch size.

Reply:
it doesn't make sense to have same checking in the until and inside the until for every iteration.
step 1: until batch_events.size == @batch_size || @events_queue.empty? do 
                  batch_events << @events_queue.pop
step 2: Once its done, now events could be either batch_size or less, either case just send to moesif.

Outer loop will check again if queue is not empty, then it will send next batch either full or until queue is empty